modules/node-shared/src/main/scala/io/constellationnetwork/node/shared/infrastructure/snapshot/managers/currency/AllowSpendOpsManager.scala

```diff
@@ -0,0 +1,222 @@
+package io.constellationnetwork.node.shared.infrastructure.snapshot.managers.currency
+
+import cats.effect.Async
+import cats.syntax.all._
+
+import scala.collection.immutable.{SortedMap, SortedSet}
+
+import io.constellationnetwork.schema._
+import io.constellationnetwork.schema.address.Address
+import io.constellationnetwork.schema.artifact.{AllowSpendExpiration, SharedArtifact, SpendTransaction}
+import io.constellationnetwork.schema.balance.{Amount, Balance, BalanceArithmeticError}
+import io.constellationnetwork.schema.epoch.EpochProgress
+import io.constellationnetwork.schema.swap._
+import io.constellationnetwork.security.Hasher
+import io.constellationnetwork.security.signature.Signed
+import io.constellationnetwork.syntax.sortedCollection.sortedSetSyntax
+
+import eu.timepit.refined.types.all.NonNegLong
+
+class AllowSpendOpsManager[F[_]: Async] {
+
+  def acceptCurrencyAllowSpends(
+    epochProgress: EpochProgress,
+    incomingCurrencyAllowSpends: SortedMap[Address, SortedSet[Signed[AllowSpend]]],
+    existentCurrencyAllowSpends: SortedMap[Address, SortedSet[Signed[AllowSpend]]],
+    allAcceptedSpendTxns: List[SpendTransaction],
+    lastUnsyncGlobalSnapshotOrdinal: SnapshotOrdinal,
+    fixingAllowSpendExpiration: SnapshotOrdinal
+  )(implicit hasher: Hasher[F]): F[SortedMap[Address, SortedSet[Signed[AllowSpend]]]] = {
+    val allAcceptedSpendTxnsAllowSpendsRefs =
+      allAcceptedSpendTxns
+        .flatMap(_.allowSpendRef)
+
+    for {
+      expiredAllowSpends <- filterExpiredAllowSpends(
+        existentCurrencyAllowSpends,
+        epochProgress,
+        allAcceptedSpendTxns,
+        lastUnsyncGlobalSnapshotOrdinal,
+        fixingAllowSpendExpiration
+      )
+
+      unexpiredAllowSpends = (incomingCurrencyAllowSpends |+| expiredAllowSpends).foldLeft(existentCurrencyAllowSpends) {
+        case (acc, (address, allowSpends)) =>
+          val lastAddressAllowSpends = acc.getOrElse(address, SortedSet.empty[Signed[AllowSpend]])
+          val unexpired = (lastAddressAllowSpends ++ allowSpends).filter(_.value.lastValidEpochProgress >= epochProgress)
+          acc + (address -> unexpired)
+      }
+
+      result <- unexpiredAllowSpends.toList.foldLeftM(unexpiredAllowSpends) {
+        case (acc, (address, allowSpends)) =>
+          allowSpends.toList.traverse(_.toHashed).map { hashedAllowSpends =>
```

Comment:
This isn't necessary to change, but an example of potential duplicate hashing here.